### DISCUSSION TITLE
Custom exception types via `builtin_exception`?

### DISCUSSION THREAD

**Discussion by inducer** [wjakob/nanobind#18] [Q&A] [answered]
Since there doesn't seem to be a mechanism for exposing custom exception types, so far I'm hackering around that by inheriting from `builtin_exception` and overriding `set_error`. (That at least compiles, haven't actually tested it yet.) On a scale of 1 to 5, how bad should I be feeling about that approach?

**Comment by wjakob** [comment, marked as answer]
The best thing would be to port pybind11's exception translators to nanobind. The basic infrastructure is already in place: the `internals` data structure defined in `src/internals.{h/cpp}` provides a field

```
    std::vector<void (*)(const std::exception_ptr &)> exception_translators;
```
which is exactly the same in pybind11. One would simply have to hook this up to "user" (i.e., developer) code so that extra translators can be appended to the list.

**Reply by inducer** [reply]
#71 has some drafty work in this direction.
